imageio/plugins/spe.py

```diff
@@ -533,37 +381,47 @@ def extract_metadata(meta: Mapping, char_encoding: str = "latin1"):
         meta.pop("exposure_sec")
 
 
-class SpeFormat(Format):
-    """See :mod:`imageio.plugins.spe`"""
+class SpePlugin(PluginV3):
+    def __init__(self, request: Request, check_filesize: bool = True) -> None:
+        """Instantiate a new SPE file plugin object
+
+        Parameters
+        ----------
+        request
+            A request object representing the resource to be operated on.
+        check_filesize
+            The number of frames in the file is stored in the file header.
+            However, this number may be wrong for certain software. If this is
+            `True` (default), derive the number of frames also from the file
+            size and raise a warning if the two values do not match.
 
-    def _can_read(self, request):
-        return request.extension in self.extensions
+        """
 
-    def _can_write(self, request):
-        return False
+        super().__init__(request)
+        if request.mode.io_mode == IOMode.write:
+            raise InitializationError("cannot write SPE files")
 
-    class Reader(Format.Reader):
-        def _open(self, char_encoding="latin1", check_filesize=True, sdt_meta=True):
-            self._file = self.request.get_file()
-            self._char_encoding = char_encoding
+        self._file = self.request.get_file()
 
-            info = self._parse_header(Spec.basic)
+        try:
+            # Spec.basic contains no string, no need to worry about character
+            # encoding.
+            info = self._parse_header(Spec.basic, "latin1")
             self._file_header_ver = info["file_header_ver"]
             self._dtype = Spec.dtypes[info["datatype"]]
             self._shape = (info["ydim"], info["xdim"])
             self._len = info["NumFrames"]
-            self._sdt_meta = sdt_meta
 
             if check_filesize:
                 # Some software writes incorrect `NumFrames` metadata.
                 # To determine the number of frames, check the size of the data
                 # segment -- until the end of the file for SPE<3, until the
                 # xml footer for SPE>=3.
-                data_end = (
-                    info["xml_footer_offset"]
-                    if info["file_header_ver"] >= 3
-                    else os.path.getsize(self.request.get_local_filename())
-                )
+                if info["file_header_ver"] >= 3:
+                    data_end = info["xml_footer_offset"]
+                else:
+                    self._file.seek(0, os.SEEK_END)
+                    data_end = self._file.tell()
```

Comment:
If this has an impact on performance, I'd suggest `seek()`ing the start of image data.

Reply:
> If this has an impact on performance, I'd suggest seek()ing the start of image data.

Yes, that would be the best place to seek to, especially if it is different from the end of the header.

The reason it will improve performance is that most OSes will opportunistically start buffering/reading a file to improve subsequential calls to read. Setting the pointer to the start of the image data before giving control back to the user allows the OS to do this while the user performs other tasks. It will likely only be a small boost, but since it is free why not :)

Reply:
This is now implemented.